doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,435 @@
+    PACKET_VALIDITY validity;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` structure from the ACK Manager overlaps this
+  significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in
+  the Application Data number space?
+- 0-RTT packets can morph into 1-RTT packets and this needs to be handled by
+  the packetiser.
+- the `packet_header` will become a structure / union most likely.
+
+Interactions
+------------
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+int ossl_quic_packetiser_buffer_frame(OSSL_QUIC_TX_PACKETISER *tx,
+                                      QUIC_STREAM *stream,
+                                      const OSSL_QUIC_FRAME *frame,
+                                      size_t frame_length);
+```
+
+### Stream Send Buffers
+
+Data from the stream send buffers is treated specially.  The packetiser knows
+how much space is left in each packet and it will request that amount of data
+from the stream send buffers.  The stream send buffers will return a
+constructed frame header and a pointer to the steam data and length.  A second
+call exists to allow the packetiser to know how much data is queued for a stream
+so that planning for the creation of multiple packets is possible.
+
+```c
+int ossl_quic_get_app_data(QUIC_STREAM *stream, size_t request,
+                           const OSSL_QUIC_FRAME **frame,
+                           const unsigned char **data,
+                           size_t *data_len);
+
+size_t ossl_quic_get_app_data_size(QUIC_STREAM *stream);
+```
+
+#### Notes
+
+* Unclear how to best free the data after sent data was acked.
+  The data will be fragments from the buffers so the stream send buffers will
+  need to remember which fragment have been sent and which are pending and
+  only free once everything is sent:
+
+```c
+int ossl_quic_free_app_data(QUIC_STREAM *stream, void *data, size_t data_len);
+```
+
+* Also need a call to tell the stream send buffers to forget about previously
+  requested app data because it needs to be retransmitted and the boundaries
+  could change.  This is similar to the free call but the data is placed/queue
+  back in the send buffers appropriately, hence it might need the stream offset as
+  well:
+
+```c
+int ossl_quic_retransmitting_app_data(QUIC_STREAM *stream, size_t offset,
+                                      void *data, size_t data_len);
```

Comment:
I'm not clear how this is interpreted. IIUC this is telling a stream send buffer that data we previously requested to be sent now needs to be retransmitted - so remove any record of the fact that we sent it before and make the data eligible to be sent again. Is that correct? What does `offset` mean here?

Reply:
`offset` is the data offset in the stream.  I change _stream offset_ to something more verbose in the description.

I'm not convinced that `offset` is required but I included it for the moment.

Your interpretation is correct.  I've reworded it a little.

Reply:
I'm unconvinced about the need for `offset`. Or at least I don't understand why we need both `data` and `offset`. They both seem to be being used to identify the data that needs to be retransmitted.

Reply:
Me either.  It depends on how the send buffers maintain the information about what has been transmitted and what hasn't.  Once that's known, we'll know which is required.

Reply:
I think you should pick one (data or offset) and run with just that in this design - not have both. We can always change it later. Having both is definitely wrong.